File: hellojava/src/main/java/com/fathzer/chess/optimized/Perft.java
package com.fathzer.chess.optimized;

import java.util.Arrays;

import com.fathzer.chess.common.PerftResult;
import com.fathzer.chess.common.PerftType;

/**
 * <a href="https://www.chessprogramming.org/Perft">Perft, ('Performance Test')</a> is a Performance Test is a debugging function
 * that walks the move generation tree of strictly legal moves to count all the leaf nodes of a certain depth,
 * which can be compared to predetermined values and used to isolate bugs.
 */
public class Perft {
    /** Performs a non bulk Perft (Performance Test) calculation.
     * @param board The board to run the performance test on.
     * @param depth The depth to run the performance test to
     * @param whitePlaying true if white is playing, false otherwise
     * @return a non null result
     */
    public PerftResult perft(Board board, int depth, boolean whitePlaying) {
        return perft(board, depth, PerftType.NON_BULK, whitePlaying);
    }

    /**  Performs a Perft (Performance Test) calculation.
     * @param board The board to run the performance test on.
     * @param depth The depth to run the performance test to
     * @param type The type of Perft to run
     * @param whitePlaying true if white is playing, false otherwise
     * @return a non null result
     */
    public PerftResult perft(Board board, int depth, PerftType type, boolean whitePlaying) {
        if (board==null) {
            throw new IllegalArgumentException("Board cannot be null");
        }
    	if (depth<=0) {
    		throw new IllegalArgumentException("Depth must be greater than 0");
    	}
        final PerftResult result = new PerftResult();
        final IntList[] moveListCache = new IntList[depth+1];
        Arrays.setAll(moveListCache, i -> new IntList());
        result.setLeafNodesCount(perft(moveListCache, board, result, depth, depth, type, whitePlaying));
        return result;
    }

    private long perft(IntList[] moveListCache, Board board, PerftResult result, int depth, int originalDepth, PerftType type, boolean whitePlaying) {
        result.incrementSearchedNodesCount();
        final IntList moves = moveListCache[depth];
        board.getMoves(moves, whitePlaying);
        if (depth == 1 && type == PerftType.NON_BULK) {
            return moves.size();
        } else if (depth == 0) {
            return 1;
        }
        long leafNodesCount = 0;
        for (int moveIndex = 0; moveIndex < moves.size(); moveIndex++) {
            int move = moves.get(moveIndex);
            board.makeMove(move);
            long moveCount = perft(moveListCache, board, result, depth - 1, originalDepth, type, !whitePlaying);
            leafNodesCount += moveCount;
            board.unmakeMove();
        }
        return leafNodesCount;
    }
}

File: hellogo/pkg/chess/optimized/perft.go
package optimized

import (
	chesscommon "hellogo/pkg/chess/common"
)

// Perft performs a Performance Test that walks the move generation tree of strictly legal moves
// to count all the leaf nodes of a certain depth, which can be compared to predetermined values
// and used to isolate bugs.
type Perft struct{}

// NewPerft creates a new Perft instance
func NewPerft() *Perft {
	return &Perft{}
}

// Perft performs a non-bulk Perft (Performance Test) calculation.
// board: The board to run the performance test on.
// depth: The depth to run the performance test to
// whitePlaying: true if white is playing, false otherwise
// Returns a non-null result
func (p *Perft) Perft(board *Board, depth int, whitePlaying bool) *chesscommon.PerftResult {
	return p.perftWithType(board, depth, chesscommon.NonBulk, whitePlaying)
}

// PerftWithType performs a Perft (Performance Test) calculation.
// board: The board to run the performance test on.
// depth: The depth to run the performance test to
// perftType: The type of Perft to run
// whitePlaying: true if white is playing, false otherwise
// Returns a non-null result
func (p *Perft) PerftWithType(board *Board, depth int, perftType chesscommon.PerftType, whitePlaying bool) *chesscommon.PerftResult {
	return p.perftWithType(board, depth, perftType, whitePlaying)
}

func (p *Perft) perftWithType(board *Board, depth int, perftType chesscommon.PerftType, whitePlaying bool) *chesscommon.PerftResult {
	if board == nil {
		panic("Board cannot be nil")
	}
	if depth <= 0 {
		panic("Depth must be greater than 0")
	}

	result := chesscommon.NewPerftResult()
	moveListCache := make([]*IntList, depth+1)
	for i := range moveListCache {
		moveListCache[i] = NewIntList()
	}

	leafNodes := p.perft(moveListCache, board, result, depth, depth, perftType, whitePlaying)
	result.SetLeafNodesCount(leafNodes)
	return result
}

func (p *Perft) perft(moveListCache []*IntList, board *Board, result *chesscommon.PerftResult, depth, originalDepth int, perftType chesscommon.PerftType, whitePlaying bool) int64 {
	result.IncrementSearchedNodesCount()
	moves := moveListCache[depth]
	board.GetMoves(moves, whitePlaying)

	if depth == 1 && perftType == chesscommon.NonBulk {
		return int64(moves.Size())
	} else if depth == 0 {
		return 1
	}

	var leafNodesCount int64
	for i := 0; i < moves.Size(); i++ {
		move := moves.Get(i)
		board.MakeMove(move)
		moveCount := p.perft(moveListCache, board, result, depth-1, originalDepth, perftType, !whitePlaying)
		leafNodesCount += moveCount
		board.UnmakeMove(move)
	}
	return leafNodesCount
}
